File: apollo/src/main/java/io/muun/apollo/domain/errors/NotificationProcessingError.java
package io.muun.apollo.domain.errors;


import io.muun.common.api.beam.notification.NotificationJson;
import io.muun.common.exception.PotentialBug;

public class NotificationProcessingError extends RuntimeException implements PotentialBug {

    /**
     *  Create an Error caused by a gap in the notification processing ID sequence.
     */
    public static NotificationProcessingError fromMissingIds(NotificationJson notification,
                                                             long lastProcessedId) {

        final String message = String.format(
                "Attempted to process notification ID=%s prevID=%s, but last processed ID was %s",
                Long.toString(notification.id),
                Long.toString(notification.previousId),
                Long.toString(lastProcessedId)
        );

        return fromCause(notification, new IllegalArgumentException(message));
    }

    /**
     * Create an Error caused by a lower-level problem during notification processing.
     */
    public static NotificationProcessingError fromCause(NotificationJson notification,
                                                        Throwable cause) {

        final String message = String.format(
                "DANGER! ERROR while processing notification %s of type %s sent by %s",
                Long.toString(notification.id),
                notification.messageType,
                notification.senderSessionUuid
        );

        return new NotificationProcessingError(message, cause);
    }

    public NotificationProcessingError(String message, Throwable cause) {
        super(message, cause);
    }
}

File: android/apollo/src/main/java/io/muun/apollo/domain/errors/notifications/NotificationProcessingError.kt
package io.muun.apollo.domain.errors.notifications


import io.muun.apollo.domain.errors.MuunError
import io.muun.common.api.beam.notification.NotificationJson
import io.muun.common.exception.PotentialBug

class NotificationProcessingError : MuunError, PotentialBug {

    companion object {

        /** Create an Error caused by a gap in the notification processing ID sequence. */
        @JvmStatic
        fun fromMissingIds(notification: NotificationJson, lastProcessedId: Long) =
            NotificationProcessingError()
                .apply {
                    addNotificationMetadata(notification)
                    metadata["lastProcessedId"] = lastProcessedId
                }

        /** Create an Error caused by a lower-level problem during notification processing. */
        @JvmStatic
        fun fromCause(notification: NotificationJson, cause: Throwable) =
            NotificationProcessingError(cause)
                .apply {
                    addNotificationMetadata(notification)
                }
    }

    private fun addNotificationMetadata(notification: NotificationJson) {
        metadata["notificationId"] = notification.id
        metadata["notificationPreviousId"] = notification.previousId
        metadata["messageType"] = notification.messageType
        metadata["senderSessionUuid"] = notification.senderSessionUuid
    }

    private constructor() :
        super("DANGER! ERROR while processing notification")

    private constructor(cause: Throwable) :
        super("DANGER! ERROR while processing notification", cause)
}
